clippy_lints/src/eta_reduction.rs

```diff
@@ -179,6 +180,26 @@ fn check_closure<'tcx>(cx: &LateContext<'tcx>, outer_receiver: Option<&Expr<'tcx
                     }
                 },
             };
+            let callee_kind = || {
+                // this lint used to refer to everything as a "function", so we add better names for particular
+                // cases, but still fall back to "function"
+                match callee_ty_adjusted.kind() {
+                    ty::FnDef(def, _) => match cx.tcx.def_kind(def) {
+                        DefKind::Ctor(CtorOf::Struct, _) => "tuple struct constructor",
+                        DefKind::Ctor(CtorOf::Variant, _) => "enum variant constructor",
+                        _ => "function",
+                    },
+                    #[expect(clippy::match_same_arms, reason = "this is not the fallback")]
+                    ty::FnPtr(..) => "function",
+                    #[expect(clippy::match_same_arms, reason = "this has a comment")]
+                    ty::Closure(..) => {
+                        // referring to this as a "closure" would've been more correct of course, but that would make
+                        // the final message read "replace the closure with the closure itself", which is.. not great
+                        "function"
+                    },
+                    _ => "function",
+                }
+            };
```

Comment:
Isn't the following nice enough?


```suggestion
            let callee_kind = || {
                match callee_ty_adjusted.kind() {
                    ty::FnDef(def, _) => cx.tcx.def_descr(*def),
                    _ => "function",
                }
            };
```

You'll get "tuple struct" instead of "tuple struct constructor" and "tuple variant" instead of "enum variant constructor", but this will better match the compiler's own naming.

And why make it a closure when it will be used only once, instead of doing this where it is required?